neqo-transport/src/connection.rs

```diff
@@ -736,89 +735,110 @@ impl Connection {
-            };
-            dump_packet(self, "<- RX", &hdr, &body);
-
-            // TODO(ekr@rtfm.com): Have the server blow away the initial
-            // crypto state if this fails? Otherwise, we will get a panic
-            // on the assert for doesn't exist.
-            // OK, we have a valid packet.
-            self.idle_timeout.on_packet_received(now);
-
-            // TODO(ekr@rtfm.com): Filter for valid for this epoch.
-
-            let ack_eliciting = self.input_packet(hdr.epoch, Decoder::from(&body[..]), now)?;
-            let space = PNSpace::from(hdr.epoch);
-            if self.acks[space].is_duplicate(hdr.pn) {
-                qdebug!([self] "Received duplicate packet epoch={} pn={}", hdr.epoch, hdr.pn);
-                self.stats.dups_rx += 1;
-                continue;
             }
-            self.acks[space].set_received(now, hdr.pn, ack_eliciting);
+            self.start_handshake(hdr, &d)?;
+            self.process_migrations(&d)?;
+        }
+        Ok(())
+    }
 
-            if matches!(self.state, State::WaitInitial) {
-                if self.role == Role::Server {
-                    assert!(matches!(hdr.tipe, PacketType::Initial(..)));
-                    // A server needs to accept the client's selected CID during the handshake.
-                    self.valid_cids.push(hdr.dcid.clone());
-                    // Install a path.
-                    assert!(self.paths.is_none());
-                    let mut p = Path::new(&d, hdr.scid.unwrap());
-                    p.local_cids.push(ConnectionId::generate(CID_LENGTH));
-                    self.paths = Some(p);
-
-                    // SecretAgentPreinfo::early_data() always returns false for a server,
-                    // but a non-zero maximum tells us if we are accepting 0-RTT.
-                    self.zero_rtt_state = if self.crypto.tls.preinfo()?.max_early_data() > 0 {
-                        ZeroRttState::Accepted
-                    } else {
-                        ZeroRttState::Rejected
-                    };
-                } else {
-                    qdebug!([self] "Changing to use Server CID={}", hdr.scid.as_ref().unwrap());
-                    let p = self
-                        .paths
-                        .iter_mut()
-                        .find(|p| p.received_on(&d))
-                        .expect("should have a path for sending Initial");
-                    p.remote_cid = hdr.scid.unwrap();
+    fn decrypt_body(&mut self, mut hdr: &mut PacketHdr, slc: &[u8]) -> Option<Vec<u8>> {
+        // Decryption failure, or not having keys is not fatal.
+        // If the state isn't available, or we can't decrypt the packet, drop
+        // the rest of the datagram on the floor, but don't generate an error.
+        let largest_acknowledged = self
+            .loss_recovery
+            .largest_acknowledged(PNSpace::from(hdr.epoch));
+        match self.crypto.obtain_crypto_state(self.role, hdr.epoch) {
```

Comment:
Is this a case where we can use `if let (...) = ...`  or even `.map`?  In other words, will clippy::pedantic immediately object to this?

Reply:
I guess we'll find out.

Reply:
Oh I'm sorry I havent seen your comment :( (my notification setup must be messed up)
We can (and should) use a map.
I tried to setup clippy::pedantic just for this function but it's complaining about a lot of other issues occuring inside it.
I hope I'll have more time to keep on refactoring towards clippy::pedantic soon!